ci: automate

CharlieC3 opened the pull request:
Closes https://github.com/hirosystems/clarity-repl/issues/28

Added a CI workflow to perform the following:
* Audits and format checks
* Build the clarity-repl and wasm on PRs
* Create a GH release and git tag on the `main` branch when appropriate (following @[conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary))
* Upload the clarity-repl WASM artifact to the GH release
* Publish the clarity-repl WASM NPM package
* Publish clarity-repl crate to creates.io

Also added a commit hook to check for commit messages which fit the conventional commit style.

No tests run for CI yet since there are none, but this step can be added once tests are merged in to `main`.

lgalabru reviewed (approved):
LGTM! So should we merge this PR https://github.com/hirosystems/clarity-repl/pull/30 from @obycode to get a few tests, and then merge this PR?
Concerning the development flow, are we moving to a branching scheme similar to clarinet with a `develop` branch? 
Thank you so much for your help on this!

CharlieC3 commented:
@lgalabru If this repo is going to be receiving more traffic, then I'd say yes a dev branch should be used. If you agree, then you can create the `dev` branch from `main`, merge #30 into `dev`, and I can update my branch off `dev` and add test runs to the CI before merging.

obycode commented:
I assume you mean that we will be merging things into the dev branch during development, and only releases get merged into main? That sounds good to me.

obycode commented:
@CharlieC3 - I created a `develop` branch (to match clarinet) and changed my PR to commit into that branch. Should we set the default branch to `develop` so that PRs open to that branch by default?

CharlieC3 commented:
@obycode Yup, good idea. I updated the default branch and created branch protection rules for `main` and `develop. I also updated this PR base branch to `develop`. Once #30 is merged, I'll account for the added tests in this PR.

obycode commented:
@lgalabru, I can take that task.

obycode commented:
Ok, those should be fixed. @lgalabru, could you just take a quick look at my commit, 273e956, to make sure those changes make sense?

lgalabru reviewed (approved):
LGTM!

CharlieC3 commented:
Thanks for fixing that, @obycode!

blockstack-devops commented:
:tada: This PR is included in version 0.17.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/clarity-repl/v/0.17.0)
- [GitHub release](https://github.com/hirosystems/clarity-repl/releases/tag/v0.17.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
